File: java-backend/src/main/java/com/shopping/cart/challange/controller/CartController.java
package com.shopping.cart.challange.controller;

import com.shopping.cart.challange.dto.AddItemDTO;
import com.shopping.cart.challange.dto.CartResponseDTO;
import com.shopping.cart.challange.service.CartService;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/cart")
public class CartController {

    private final CartService cartService;


    public CartController(CartService cartService) {
        this.cartService = cartService;
    }

    /**
     * Adds an item to the cart.
     *
     * @param addItemDTO the AddItemDTO containing the item name
     * @return ResponseEntity with a success or error message
     */
    @PostMapping("/add")
    public ResponseEntity<String> addItem(@RequestBody AddItemDTO addItemDTO) {
        try {
            if (addItemDTO == null) {
                return ResponseEntity.badRequest().body("Pass a Valid add Item Payload");
            }
            if (addItemDTO.getItemName() == null || addItemDTO.getItemName().trim().isEmpty()) {
                return ResponseEntity.badRequest().body("Item name cannot be null or empty");
            }
            String result = cartService.addItem(addItemDTO.getItemName());
            return ResponseEntity.ok(result);
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body("Invalid item: '" + addItemDTO.getItemName() + "' is not a valid item.");
        } catch (Exception e) {
            return ResponseEntity.internalServerError().body("Error adding item: " + e.getMessage());
        }
    }

    /**
     * Retrieves all items in the cart with their quantities and total amount.
     *
     * @return ResponseEntity containing the CartResponseDTO
     */
    @GetMapping("/getAllItems")
    public ResponseEntity<CartResponseDTO> getCartItems() {
        try {
            CartResponseDTO cartResponse = cartService.getAllItems();
            return ResponseEntity.ok(cartResponse);
        } catch (Exception e) {
            return ResponseEntity.internalServerError().body(null);
        }
    }

    /**
     * Fetches the total cost of the cart.
     *
     * @return ResponseEntity with the total cost
     */
    @GetMapping("/getTotal")
    public ResponseEntity<Double> getTotalCost() {
        try {
            double totalCost = cartService.fetchTotalCost();
            return ResponseEntity.ok(totalCost);
        } catch (Exception e) {
            return ResponseEntity.internalServerError().body(null);
        }
    }
}

File: js-proxy/src/controllers/cartController.js
// js-api/src/controllers/cartController.js
const axios = require('axios');

// Base URL of the Java backend
const JAVA_BACKEND_URL = process.env.JAVA_BACKEND_URL || 'http://localhost:8080';

// Add an item to the cart
const addItem = async (req, res, next) => {
  try {
    const response = await axios.post(`${JAVA_BACKEND_URL}/api/cart/add`, req.body);
    res.status(200).json({ message: response.data });
  } catch (error) {
    next(error); // Pass to error handler
  }
};

// Get all items in the cart
const getAllItems = async (req, res, next) => {
  try {
    const response = await axios.get(`${JAVA_BACKEND_URL}/api/cart/getAllItems`);
    res.status(200).json(response.data);
  } catch (error) {
    next(error);
  }
};

// Get the total cost of the cart
const getTotalCost = async (req, res, next) => {
  try {
    const response = await axios.get(`${JAVA_BACKEND_URL}/api/cart/getTotal`);
    res.status(200).json(response.data);
  } catch (error) {
    next(error);
  }
};

module.exports = {
  addItem,
  getAllItems,
  getTotalCost,
};